nucypher/blockchain/economics.py

```diff
@@ -14,60 +14,166 @@
 You should have received a copy of the GNU Affero General Public License
 along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
 """
+from typing import Tuple
+
+import math
+from decimal import Decimal
+
 from math import log
 
 
-class Economics:
-    def __init__(
-            self,
-            initial_supply=10**9,
-            initial_inflation=1.0,
-            T_half=2.0,
-            T_sat=1.0,
-            small_staker_multiplier=0.5) -> None:
+class TokenEconomics:
+    """
+    Calculate parameters to use in token and escrow deployment
+    from high-level human-understandable parameters.
+
+    Formula for staking in one period:
+    (totalSupply - currentSupply) * (lockedValue / totalLockedValue) * (k1 + allLockedPeriods) / k2
+
+    K2 - Staking coefficient
+    K1 - Locked periods coefficient
+
+    if allLockedPeriods > awarded_periods then allLockedPeriods = awarded_periods
+    kappa * log(2) / halving_delay === (k1 + allLockedPeriods) / k2
+
+    ...but also...
+
+    kappa = (small_stake_multiplier + (1 - small_stake_multiplier) * min(T, T1) / T1)
+    where allLockedPeriods == min(T, T1)
+    """
+
+    # Token Denomination
+    __token_decimals = 18
+    nunits_per_token = 10 ** __token_decimals          # Smallest unit designation
+
+    # Period Definition
+    hours_per_period = 24                            # Hours in single period
+    seconds_per_period = hours_per_period * 60 * 60  # Seconds in single period
+
+    # Time Constraints
+    minimum_locked_periods = 30                      # 720 Hours minimum
+
+    # Value Constraints
+    minimum_allowed_locked = 15000 * nunits_per_token
+    maximum_allowed_locked = int(4e6) * nunits_per_token
+
+    # Supply
+    __default_initial_supply = Decimal(int(1e9) * nunits_per_token)
```

Comment:
I actually think that these can be int. And we'll have fully int once we passed `__init__` (because in solidity it's all `uint256`)

Reply:
:heavy_check_mark: